src/operation/valid/RepeatedPointRemover.cpp

```diff
@@ -38,16 +38,49 @@ RepeatedPointRemover::removeRepeatedPoints(const geom::CoordinateSequence* seq)
     pts->push_back(*prevPt) ;
 
     for (std::size_t i = 1; i < sz; i++) {
-        const Coordinate* nextPt = &(seq->getAt(i));
-        if (*nextPt != *prevPt) {
-            pts->push_back(*nextPt);
+        const Coordinate* curPt = &(seq->getAt(i));
+        if (*curPt != *prevPt) {
+            pts->push_back(*curPt);
+            prevPt = curPt;
         }
-        prevPt = nextPt;
     }
 
     return detail::make_unique<geom::CoordinateArraySequence>(pts.release(), seq->getDimension());
 }
 
+std::unique_ptr<geom::CoordinateArraySequence>
+RepeatedPointRemover::removeRepeatedAndInvalidPoints(const geom::CoordinateSequence* seq) {
+    using geom::Coordinate;
+    std::size_t start = 0;
+
+    if (seq->isEmpty()) {
+        return detail::make_unique<geom::CoordinateArraySequence>(0u, seq->getDimension());
+    }
+
+    auto pts = detail::make_unique<std::vector<Coordinate>>();
+    auto sz = seq->getSize();
+    pts->reserve(sz); // assume not many points are repeated
+
+    // Find first valid poit
+    while(!seq->getAt(start).isValid()) {
+        start++;
+    }
+
+    const Coordinate* prevPt = &(seq->getAt(start));
```

Comment:
Implementing this as a `CoordinateFilter` will be faster, if that matters

Reply:
Probably would be generally beneficial, RepeatedPointRemover shows up in a few other places.

Reply:
It would be very similar to https://github.com/libgeos/geos/blob/main/src/index/chain/MonotoneChainBuilder.cpp#L49